File: VDFEditor/src/main/java/de/treichels/hott/vdfeditor/ui/ExceptionDialog.java
package de.treichels.hott.vdfeditor.ui;

import java.io.PrintWriter;
import java.io.StringWriter;

import gde.util.Util;
import javafx.application.Platform;
import javafx.scene.control.Alert;
import javafx.scene.control.Label;
import javafx.scene.control.TextArea;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Priority;
import javafx.stage.Stage;

public class ExceptionDialog extends Alert {
    private static boolean SHOWING = false;

    synchronized static void show(final Throwable throwable) {
        // show only one instance of the dialog at a time
        if (!SHOWING) {
            SHOWING = true;
            Platform.runLater(() -> {
                new ExceptionDialog(throwable).showAndWait();
                SHOWING = false;
            });
        }
    }

    private ExceptionDialog(final Throwable throwable) {
        super(AlertType.ERROR);

        if (Util.DEBUG) throwable.printStackTrace();

        ((Stage) getDialogPane().getScene().getWindow()).getIcons().add(Controller.ICON);
        setHeaderText(null);

        String message = throwable.getLocalizedMessage();
        if (message == null) message = throwable.getMessage();
        if (message == null) message = throwable.getClass().getSimpleName();
        setContentText(message);

        // Create expandable Exception.
        final StringWriter sw = new StringWriter();
        final PrintWriter pw = new PrintWriter(sw);
        throwable.printStackTrace(pw);
        final String exceptionText = sw.toString();

        final Label label = new Label("The exception stacktrace was:");

        final TextArea textArea = new TextArea(exceptionText);
        textArea.setEditable(false);
        textArea.setWrapText(true);

        textArea.setMaxWidth(Double.MAX_VALUE);
        textArea.setMaxHeight(Double.MAX_VALUE);
        GridPane.setVgrow(textArea, Priority.ALWAYS);
        GridPane.setHgrow(textArea, Priority.ALWAYS);

        final GridPane expContent = new GridPane();
        expContent.setMaxWidth(Double.MAX_VALUE);
        expContent.add(label, 0, 0);
        expContent.add(textArea, 0, 1);

        // Set expandable Exception into the dialog pane.
        getDialogPane().setExpandableContent(expContent);
    }
}

File: HoTT-Model/src/main/java/de/treichels/hott/util/ExceptionDialog.kt
package de.treichels.hott.util

import java.io.PrintWriter
import java.io.StringWriter

import de.treichels.hott.messages.Messages
import javafx.application.Platform
import javafx.scene.control.Alert
import javafx.scene.image.Image
import javafx.scene.layout.Priority
import javafx.stage.Stage
import tornadofx.*

class ExceptionDialog(throwable: Throwable) : Alert(Alert.AlertType.ERROR) {
    init {
        if (Util.DEBUG) throwable.printStackTrace()

        title = Messages.getString("Error")
        (dialogPane.scene.window as Stage).icons += Image(ExceptionDialog::class.java.getResourceAsStream("/icon.png"))
        headerText = null

        var message: String? = throwable.localizedMessage
        if (message == null) message = throwable.message
        if (message == null) message = throwable.javaClass.simpleName
        contentText = message

        // Create expandable Exception.
        val sw = StringWriter()
        throwable.printStackTrace(PrintWriter(sw))

        // Set expandable Exception into the dialog pane.
        dialogPane.expandableContent = gridpane {
            maxWidth = Double.MAX_VALUE
            label("The exception stacktrace was:") {
                gridpaneConstraints { columnRowIndex(0,0) }
            }
            textarea(sw.toString()) {
                isEditable = false
                isWrapText = true
                maxWidth = Double.MAX_VALUE
                maxHeight = Double.MAX_VALUE
                gridpaneConstraints {
                    columnRowIndex(0,1)
                    vhGrow = Priority.ALWAYS
                }
            }
        }
    }

    companion object {
        private var SHOWING = false

        @Synchronized fun show(throwable: Throwable) {
            // show only one instance of the dialog at a time
            if (!SHOWING) {
                SHOWING = true
                Platform.runLater {
                    ExceptionDialog(throwable).showAndWait()
                    SHOWING = false
                }
            }
        }
    }
}
